Fixes composer installation errors due to problematic configuration files

weierophinney opened the pull request:
If a module configuration file contains elements that depend on autoloading and/or elements that are clearly not configuration, calling `include` on it can lead to failure to complete a composer installation/removal action.

This patch modifies the logic as follows:

- If a configuration file does not have `asset_manager` configuration, it is skipped.
- Both installation and uninstallation actions tokenize the configuration file to look for problematic elements, notably `exec()` and `eval()` calls; these raise warnings by the installer if detected.
- Instead of performing install operations immediately, the `Plugin` class now aggregates install operations as closures around the relevant events, and executes these during a `post-autoload-dump` plugin hook; this fixes autoloading issues. Uninstall operations still occur immediately as the relevant uninstall and update events.

Fixes #4

weierophinney committed:
Fixes composer installation errors due to problematic configuration files

If a module configuration file contains elements that depend on
autoloading and/or elements that are clearly not configuration, calling
`include` on it can lead to failure to complete a composer
installation/removal action.

This patch modifies the logic to tokenize the configuration file and
look for problematic elements; when encountered, the (un)installer will skip
including that configuration file, and instead emit a warning.

weierophinney committed:
Fix Travis configuration to ensure deps appropriate to PHP version are installed

michalbundyra commented on .travis.yml:
```diff
@@ -47,18 +48,22 @@ matrix:
 before_install:
-  - if [[ $TRAVIS_PHP_VERSION != "hhvm" && $TEST_COVERAGE != 'true' ]]; then phpenv config-rm xdebug.ini ; fi
+  - if [[ $TEST_COVERAGE != 'true' ]]; then phpenv config-rm xdebug.ini || return 0 ; fi
+  - travis_retry composer self-update
```
This line is redundant. Travis update composer automatically, please see:
https://travis-ci.org/zfcampus/zf-asset-manager/jobs/296348988#L473
https://travis-ci.org/zfcampus/zf-asset-manager/jobs/296348988#L478

weierophinney committed:
Do not emit warnings for class pseudo-constant or files without asset_manager configuration

If a configuration file does not contain `asset_manager` configuration,
we do not need to do anything with it in the first place, and do not
need to emit warnings.

If the file contains any class pseudo-constant usage (e.g.
`AssetInstallerFactory::class`), we do not need to flag the file as
unusable, as PHP will resolve these regardless of autoloading.

As such, we'll only have issues with class constants, static members,
and cases where we are calling `new` or extending or implementing
classes or interfaces.

michalbundyra commented on src/UnparseableTokensTrait.php:
```diff
@@ -0,0 +1,82 @@
+    ];
+
+    /**
+     * @var string $packageConfigPath
```
`@param` instead of `@var`

michalbundyra commented on src/UnparseableTokensTrait.php:
```diff
@@ -0,0 +1,82 @@
+    }
+
+    /**
+     * @var string $packageConfigPath
```
As above, `@param` instead of `@var`

weierophinney committed:
Only consider exit and eval statements as unparseable

weierophinney committed:
Modifies plugin to delay install/uninstall operations until autoload dumping

Per conversations with @Xerkus, this patch modifies the `Plugin` class
to create closures around each install, uninstall, and update event in
order to delay their execution until the `post-autoload-dump` event
executes. What that event executes, it will:

- loop through all uninstall closures and execute them.
- loop through all install closures and execute them.

This approach should solve autoloading issues (e.g., when resolving
class constants).

weierophinney committed:
Fixes parameter annotations in UnparseableTokensTrait

weierophinney committed:
Removes redundant `composer self-update` command from travis config

weierophinney committed:
Refactors to always run uninstall operations immediately

Uninstall operations should not suffer the same autoloading issues, as
the autoloader will already be in place. However, if they run during
post-autoload-dump, they _will_ have issues. As such, this commit
modifies the logic to run uninstall operations immediately, instead of
memoizing them to run later.

Xerkus commented on src/Plugin.php:
```diff
@@ -55,21 +63,35 @@ public function activate(Composer $composer, IOInterface $io)
      *
      * Uninstalls any previously installed assets for the package, and then
-     * runs an install for the package.
+     * memoizes an install operation to run post-autoload-dump.
```
typo in memorizes

weierophinney replied:
No typo: read about [memoization](https://en.wikipedia.org/wiki/Memoization) on wikipedia. :smile:

Xerkus commented on src/UnparseableTokensTrait.php:
```diff
@@ -0,0 +1,75 @@
+            }
+
+            // ::class notation is okay
+            if ($token[0] === T_DOUBLE_COLON
```
this check is no longer needed

Xerkus commented on src/UnparseableTokensTrait.php:
```diff
@@ -0,0 +1,75 @@
+     * @param string|array Token to test as a class pseudoconstant
+     * @return bool
+     */
+    private function isClassPseudoConstant($token)
```
same here, no longer needed

weierophinney committed:
Adds a pre-package-update hook

Moves the uninstall operations originally in the post-package-update
hook to a pre-package-update hook. This should autoloading stays the
same (in cases where classes, class constants, config files, etc.
disappear).

weierophinney committed:
Removed obsolete checks from UnparseableTokensTrait::isParseableContent()

weierophinney committed:
Updates to composer-plugin-api 1.1

Prevents deprecation notices.

weierophinney committed:
Ensure appropriate deps are installed based on PHP version
